ffi/src/winscard/buf_alloc.rs

```diff
@@ -1,121 +1,127 @@
-use std::iter::once;
 use std::slice::from_raw_parts_mut;
 
-use ffi_types::{LpByte, LpDword, LpStr, LpWStr};
+use ffi_types::{LpByte, LpDword, LpWStr};
 use winscard::{Error, ErrorKind, WinScardResult};
 
-use super::scard_handle::WinScardContextHandle;
-use crate::utils::str_to_w_buff;
+use super::scard_handle::{OutBuffer, RequestedBufferType};
 
 pub const SCARD_AUTOALLOCATE: u32 = 0xffffffff;
 
-pub unsafe fn copy_buff(
-    context: &mut WinScardContextHandle,
-    raw_buff: LpByte,
-    raw_buff_len: LpDword,
-    buff_to_copy: &[u8],
-) -> WinScardResult<()> {
-    let buff_to_copy_len = buff_to_copy.len().try_into()?;
-
-    if raw_buff.is_null() {
-        unsafe {
-            *raw_buff_len = buff_to_copy_len;
-        }
-        return Ok(());
+/// This function decides how to treat the provided buffer by the user and how to return the requested data.
+///
+/// When the user requests some data from the smart card using the WinSCard API, we have three ways
+/// how to handle it:
+/// * write data in the provided buffer.
+/// * write data length of the requested data in the provided length pointer.
+/// * allocate data by ourselves and write data pointer in the provided buffer.
+pub unsafe fn build_buf_request_type<'data>(
+    p_buf: LpByte,
+    pcb_buf: LpDword,
+) -> WinScardResult<RequestedBufferType<'data>> {
+    if pcb_buf.is_null() {
+        return Err(Error::new(ErrorKind::InvalidParameter, "pcb_buf cannot be null"));
     }
 
-    if unsafe { *raw_buff_len } == SCARD_AUTOALLOCATE {
-        // Allocate a new buffer and write an address into raw_buff.
-        let allocated = context.allocate_buffer(buff_to_copy.len())?;
-        unsafe {
-            *(raw_buff as *mut *mut u8) = allocated;
-            *raw_buff_len = buff_to_copy_len;
-            from_raw_parts_mut(allocated, buff_to_copy.len()).copy_from_slice(buff_to_copy);
-        }
+    Ok(if p_buf.is_null() {
+        // If this value is NULL, we ignore the buffer length, writes the length of the buffer that
+        // would have been returned if this parameter had not been NULL, and returns a success code.
+        RequestedBufferType::Length
+    } else if
+    // SAFETY: The `pcb_buf` parameter cannot be null. We've checked for it above.
+    unsafe { *pcb_buf } == SCARD_AUTOALLOCATE {
+        // If the buffer length is specified as SCARD_AUTOALLOCATE, then data pointer is
+        // converted to a pointer to a byte pointer, and receives the address of a block of memory
+        // containing the attribute.
+        RequestedBufferType::Allocate
```

Comment:
style: The formatting is a bit strange. Maybe it could be improved by splitting on several lines.

Reply:
done: [`7b6cb00` (#250)](https://github.com/Devolutions/sspi-rs/pull/250/commits/7b6cb00e1b2110172d2de88e2374daa22cc8fc32)